src/parser/parser.rs

```diff
@@ -11,99 +11,130 @@ use serde_json::Value;
 #[grammar = "parser/grammar/json_path.pest"]
 struct JsonPathParser;
 
-/// the parsing function.
-/// Since the parsing can finish with error the result is [[Result]]
-#[allow(clippy::result_large_err)]
-pub fn parse_json_path(jp_str: &str) -> Result<JsonPath, Error<Rule>> {
-    Ok(parse_internal(
-        JsonPathParser::parse(Rule::path, jp_str)?.next().unwrap(),
-    ))
+/// Parses a string into a [JsonPath].
+///
+/// # Errors
+///
+/// Returns a variant of [JsonPathParserError] if the parsing operation failed.
+pub fn parse_json_path(jp_str: &str) -> Result<JsonPath, JsonPathParserError> {
+    match JsonPathParser::parse(Rule::path, jp_str)?.next() {
+        Some(parsed_pair) => Ok(parse_internal(parsed_pair)?),
+        None => Err(JsonPathParserError::ParserError(format!(
+            "Failed to parse JSONPath {jp_str}"
+        ))),
+    }
 }
 
 /// Internal function takes care of the logic by parsing the operators and unrolling the string into the final result.
-fn parse_internal(rule: Pair<Rule>) -> JsonPath {
+///
+/// # Errors
+///
+/// Returns a variant of [JsonPathParserError] if the parsing operation failed
+fn parse_internal(rule: Pair<Rule>) -> Result<JsonPath, JsonPathParserError> {
     match rule.as_rule() {
         Rule::path => rule
             .into_inner()
             .next()
-            .map(parse_internal)
-            .unwrap_or(JsonPath::Empty),
-        Rule::current => JsonPath::Current(Box::new(
+            .ok_or(JsonPathParserError::ParserError(
+                "expected valid Rule::path token but found found nothing".to_string(),
+            ))
+            .and_then(parse_internal),
+        Rule::current => Ok(JsonPath::Current(Box::new(
             rule.into_inner()
                 .next()
                 .map(parse_internal)
-                .unwrap_or(JsonPath::Empty),
-        )),
-        Rule::chain => JsonPath::Chain(rule.into_inner().map(parse_internal).collect()),
-        Rule::root => JsonPath::Root,
-        Rule::wildcard => JsonPath::Wildcard,
-        Rule::descent => parse_key(down(rule))
-            .map(JsonPath::Descent)
-            .unwrap_or(JsonPath::Empty),
-        Rule::descent_w => JsonPath::DescentW,
-        Rule::function => JsonPath::Fn(Function::Length),
-        Rule::field => parse_key(down(rule))
-            .map(JsonPath::Field)
-            .unwrap_or(JsonPath::Empty),
-        Rule::index => JsonPath::Index(parse_index(rule)),
-        _ => JsonPath::Empty,
+                .unwrap_or(Ok(JsonPath::Empty))?,
```

Comment:
@besok handling the `Rule::current` variant with `ok_or(Err())` causes the following tests to fail:
```
failures:

---- parser::parser::tests::path_test stdout ----
thread 'parser::parser::tests::path_test' panicked at 'parsing error expected valid Rule::current token but found nothing', src\parser\parser.rs:273:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- parser::parser::tests::index_filter_test stdout ----
thread 'parser::parser::tests::index_filter_test' panicked at 'parsing error expected valid Rule::current token but found nothing', src\parser\parser.rs:273:17

---- tests::no_value_chain_test stdout ----
thread 'tests::no_value_chain_test' panicked at 'the path is correct: "expected valid Rule::current token but found nothing"', src\lib.rs:1141:65

---- tests::no_value_filter_from_not_arr_filter_test stdout ----
thread 'tests::no_value_filter_from_not_arr_filter_test' panicked at 'the path is correct: "expected valid Rule::current token but found nothing"', src\lib.rs:1092:68


failures:
    parser::parser::tests::index_filter_test
    parser::parser::tests::path_test
    tests::no_value_chain_test
    tests::no_value_filter_from_not_arr_filter_test
```
I'm not sure if this a problem with the tests or a problem with the logic of their not being a value when calling `Rule::current.into_inner()`, so I've left it with the previous behavior of `unwrap_or(JsonPath::Empty)` which was there before.

Reply:
Ah, yeah, okay, no problems. Let's leave it as is. I will take a look at weekend, hopefully,